test/framework/src/main/java/org/elasticsearch/snapshots/AbstractSnapshotIntegTestCase.java

```diff
@@ -366,4 +374,36 @@ protected long getCountForIndex(String indexName) {
     protected void assertDocCount(String index, long count) {
         assertEquals(getCountForIndex(index), count);
     }
+
+    protected void awaitNoMoreRunningOperations(String viaNode) throws Exception {
+        logger.info("--> verify no more operations in the cluster state");
+        awaitClusterState(viaNode, state -> state.custom(SnapshotsInProgress.TYPE, SnapshotsInProgress.EMPTY).entries().isEmpty() &&
+                state.custom(SnapshotDeletionsInProgress.TYPE, SnapshotDeletionsInProgress.EMPTY).hasDeletionsInProgress() == false);
+    }
+
+    private void awaitClusterState(String viaNode, Predicate<ClusterState> statePredicate) throws Exception {
+        final ClusterService clusterService = internalCluster().getInstance(ClusterService.class, viaNode);
+        final ThreadPool threadPool = internalCluster().getInstance(ThreadPool.class, viaNode);
+        final ClusterStateObserver observer = new ClusterStateObserver(clusterService, logger, threadPool.getThreadContext());
+        if (statePredicate.test(observer.setAndGetObservedState()) == false) {
```

Comment:
I like this a lot better than:

```java
dataNodeClient().admin().cluster().prepareState().get().getState();
```
+ busy assert.

Often times the tests use this kind of waiting when the busy assert will fail for a bit and then waste a second or two until the next run because of exponential back-off (over the large number of tests that do this kind of waiting for a certain CS it's well worth taking this approach IMO, especially with the concurrent snapshot ITs adding a large number of new tests that need this thing).
Also, the `client()` approach can (in disruption ITs) randomly go for the client of an isolated node and then waste effort and time for retrying in the transport master node action.

Reply:
I understand the motivation; I'm wondering if it should always wait for next change?

Reply:
I think waiting for the next change would make it very hard to avoid races. We often have this pattern:
1. do operation
2. wait for no more operations running

If the first step completes before we start waiting we dead-lock. And this is just one example, the concurrent snapshotting tests make use of this logic in other spots where similar races could occur

Reply:
Ok. I was wondering if something similar could happen: 1. do operation and 2. check cluster state on a data node that has not yet processed the updated cluster state

Reply:
:+1: yea that was an issue in the concurrency tests, in the end it just requires ensuring that stuff actually started by some other means before waiting for it to go away :)

Reply:
Good! Thanks